Add code-styles for controlling parentheses in a project ('coarse' approach).

CyrusNajmabadi opened the pull request:
Alternative approach to: https://github.com/dotnet/roslyn/pull/21316

This PR was borne out of the design questions raised in that PR here: https://github.com/dotnet/roslyn/pull/21316#issuecomment-379063288.  It specifically implements my preferences for how we might have a 'coarse' approach to parentheses options.  So, instead of the fine-grained approach taken in that PR, this approach limits the number of options to '3' for both C# and VB.  Specifically:

![image](https://user-images.githubusercontent.com/4564579/38459972-b43ca670-3a65-11e8-9620-93ced42266b8.png)

Both the 'in arithmetic binary operators' and 'in other binary operators' options allow the user to state if they want parentheses to be required for clarity, or removed if unnecessary:

![image](https://user-images.githubusercontent.com/4564579/38459988-0e59624c-3a66-11e8-8ec7-8c44d9f37514.png)

Note: due to the high chance for introducing ambiguous/confusing code, certain operator are exempt from the "never if unncessary" rule.  For example, we will never remove the parens from ```(1 + 2) << 3``` (even though we can), because there is high audience confusion as to what ```1 + 2 << 3``` means.

CyrusNajmabadi commented:
retest this please

CyrusNajmabadi commented:
@sharwell @jinujoe @kuhlenh Can you please advise on the design decisions here?  This PR is roughly 8.5 months old at this point, and maintaining it has not been pleasant.  I think this approach provides a very reasonable user experience that does not overwhelm with options, while also being flexible enough to handle the the vast majority of user groups we can envision wanting this feature.

CyrusNajmabadi commented:
@sharwell please let me know the sorts of things you'd like to see in the preview.

CyrusNajmabadi commented:
Note: the example i currently have is:

```
        // {ServicesVSResources.Prefer_colon}
        var v = ((a <= b) || (c && d)) == (e is f);

        // {ServicesVSResources.Over_colon}
        var v = a <= b || c && d == e is f;
```

CyrusNajmabadi commented:
Also, note that @jcouv already reviewed https://github.com/dotnet/roslyn/pull/21316.  So i think we really only need one more review here.

CyrusNajmabadi commented:
Note: i recommend reviewing https://github.com/dotnet/roslyn/pull/26523 before htis one as it will help keep this PR smaller.

CyrusNajmabadi commented:
retest this please

CyrusNajmabadi commented:
@sharwell @jinujoseph can this be assigned to someone?

CyrusNajmabadi commented:
@sharwell this would be my preferred PR to review.  Note: i think it can go in as is without a full decision on what hte best preview string is.  That string can trivially be changed at a later point in time if necessary.  Whereas, the feature still has a ton of value going in as is right now.

sharwell commented on src/VisualStudio/CSharp/Impl/Options/Formatting/StyleViewModel.cs:
```diff
@@ -876,6 +878,113 @@ class Customer2
+//[
+        // {ServicesVSResources.Prefer_colon}
+        var v = a + (b * c);
+
```
:bulb: These previews should be updated to account for #25574

sharwell commented on src/Workspaces/Core/Portable/CodeStyle/CodeStyleOptions.cs:
```diff
@@ -221,5 +221,68 @@ private static CodeStyleOption<AccessibilityModifiersRequired> ParseAccessibilit
+                nameof(CodeStyleOptions), fieldName, defaultValue,
+                storageLocations: new OptionStorageLocation[]{
+                    new EditorConfigStorageLocation<CodeStyleOption<ParenthesesPreference>>(
+                        $"dotnet_style_parentheses_in_{styleName}",
```
❕ I frequently use text searches to find the definitions of .editorconfig options. Don't use formatting in a manner that prevents the complete option name from appearing as a literal in source code.

CyrusNajmabadi replied:
sure.  this was more relevant when there were like 9 options. now that's 3, just repeating the full name is reasonably.

CyrusNajmabadi replied:
fixed.

sharwell commented on src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs:
```diff
@@ -0,0 +1,1655 @@
+{
+    void M()
+    {
+        if ( $$(a[0]) is string s) { }
```
❕ Also need tests for the following (didn't see them in a quick search):

```
if ((expression is bool b) && b) { }

if (!(expression is bool b)) { }
```

CyrusNajmabadi replied:
tests added.

sharwell commented on src/Features/CSharp/Portable/AddRequiredParentheses/CSharpAddRequiredParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,80 @@
+            => node is BinaryExpressionSyntax ||
+               node is IsPatternExpressionSyntax isPattern && isPattern.Pattern is ConstantPatternSyntax;
+
+        protected override (ExpressionSyntax, SyntaxToken, ExpressionSyntax) GetPartsOfBinaryLike(ExpressionSyntax binaryLikeOpt)
```
❕ The argument is not optional

:bulb: Named tuple elements

CyrusNajmabadi replied:
renamed.  In this case i'm ok with non-named.  i think it's very clear from context.

sharwell commented on src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs:
```diff
@@ -0,0 +1,1655 @@
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnnecessaryParentheses
+{
+    public partial class RemoveUnnecessaryParenthesesTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
```
❕ Additional test cases:

```
(this.Property) = Property;
```

```
x += (3 * 2);
```

```
x *= (3 + 2);
```

```
((short)3).ToString();
```

```
int x = 3 * (checked(5));
```

```
int x = 3 * (unchecked(5));
```

```
string property = "My " + (nameof(property));
```

```
bool x = ("" is string);
```

```
string x = ("" is string).ToString();
```

```
string y;
string x = (y = "text");
```

```
Func<string, string> y1 = (v) => v;
Func<string, string> y2 = (v => v);
Func<string, string> y3 = ((v) => v);
Func<string, string> z1 = () => "";
Func<string, string> z2 = (() => "");
```

```
string y = ((Func<string, string>)((v) => v))("text");
```

```
(await Method()).OtherMethod();
```

```
#if (A || B)
#if(A)
#if (A)
```

```
return(value);
return (value);
```

CyrusNajmabadi replied:
Note: the code we're trying to format looks like this:

```c#
    {
#ifA || B
#endif
    }
```

There is elastic trivia between the ```if``` and the ```A```.  But this doesn't actually get formatted.  I can see the formatter examining the ```{``` and ```}``` as the only pair of tokens to look at.  But i can't see it ever diving into the directives.  I'm not sure how this is supposed to work?  Can you advise?

sharwell commented:
In reviewing the Remove Unnecessary Usings feature for C# code with the default settings, I found the following situations potentially confusing. I'll update the **Expected** column based on the results of a design review with the team.

| Expression | Expected | Actual |
| --- | --- | --- |
| `(x == null) == (y == null)` | | `x == null == (y == null)` |
| `(x == null) == y` |  | `x == null == y` |
| `*(int*)(&x)` | | `*(int*)&x` |
| `(long)(-1)` | | `(long)-1` |
| `"Text " + (count++) + " suffix"` | | `"Text " + count++ + " suffix"` |
| `(x % 10000) / 1000` | | `x % 10000 / 1000` |

sharwell commented:
In reviewing the Add Parentheses for Clarity for C# code with the default settings, I found the following situations to review. Like above, I'll update the **Expected** column based on the results of a design review.

| Expression | Expected | Actual |
| --- | --- | --- |
| `x != null && x.Length > 3` | | `(x != null) && (x.Length > 3)` |

sharwell commented:
❗️ The following is a bad case

Input:

```csharp
if (0 >= 3 * 2 / 4)
{
}
```

Add parentheses Fix All result:

```csharp
if (0 >= ((3 * 2 / 4)))
{
}
```

This produces a Remove Unnecessary Parentheses diagnostic with a Fix All that restores the original input.

sharwell commented on src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs:
```diff
@@ -1651,5 +1651,26 @@ void M()
+    void M(object expression)
+    {
+        if (expression is bool b && b) { }
+    }
```
There is already a test showing the parentheses are not removed

CyrusNajmabadi replied:
> ```expression is b && b``` (constant pattern)?

That's shouldn't be a constant pattern.  That's just "foo is string && ..."  If that was a constant pattern, that would break a lot of code.

sharwell commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,121 @@
+            var additionalLocations = ImmutableArray.Create(parenthesizedExpression.GetLocation());
+
+            context.ReportDiagnostic(Diagnostic.Create(
+                CreateUnnecessaryDescriptor(severity),
```
❕ `UnnecessaryWithoutSuggestionDescriptor`

CyrusNajmabadi replied:
Fixed.

sharwell commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,121 @@
+                parenthesizedExpression.GetFirstToken().GetLocation(), additionalLocations));
+
+            context.ReportDiagnostic(Diagnostic.Create(
+                CreateUnnecessaryDescriptor(DiagnosticSeverity.Hidden),
```
❕ `UnnecessaryWithoutSuggestionDescriptor`

CyrusNajmabadi commented:
> ```(x == null) == (y == null)  ->  x == null == (y == null)```

In the top table, this is about the only one that i think isn't great.  But i'm not sure i've internalized why?  Maybe it's the break in symmetry?  Can we come up with a heuristic that indicates why that is undesirable?  i.e. "in a sequence of equal precedence operators, if a later of the sequence needs parentheses for appropriate precedence, don't remove parens from an earlier part"?

CyrusNajmabadi commented:
For your second table, this all seems reasonable to me.  After all, is it:

```(hr != 0) || !equivalent``` or is it ```hr != (0 || !equivalent)```?

Parentheses clarify things.  Note: as mentioned in our talk, this would be one of those areas where fine-grained rules address the problem succinctly since you could control this behavior instead of having it all be bucketed into the "other binary operators" group.

CyrusNajmabadi commented:
> Add parentheses Fix All result:

Fixed the fix-all issue.  Good catch.  Thanks!

CyrusNajmabadi commented:
@sharwell I have addressed all your feedback.  please let me know the conclusions you come to in the design meetings.  Note: for your tables, i'm curious what the behavior of stylecop analyzers is.

CyrusNajmabadi commented:
test windows_debug_vs-integration_prtest please

CyrusNajmabadi commented:
Sam and i discussed this more, especially in the context of this commit where he tried out the feature: https://github.com/sharwell/roslyn/commit/b8f79b78077597d6d5fed67df9eec6d248a4a60a

right now, the current split seems problematic.  Firstly, because the "other binary operators" category seems overly broad.  Many would want to turn it on because they want parens for &&/|| expressions.  However, this also pulls in all relational/equality operators.  Leading to messes like converting this:

```c#
                    if (type.SpecialType == SpecialType.System_Object ||
                        operandType.IsArray() && type.IsArray() &&
                        ((ArrayTypeSymbol)type).ElementType.SpecialType == SpecialType.System_Object)
```

Into: 

```c#
                    if (type.SpecialType == SpecialType.System_Object ||
                        (operandType.IsArray() && type.IsArray() &&
                        ((ArrayTypeSymbol)type).ElementType.SpecialType == SpecialType.System_Object))
```

We discussed if this could be resolved by breaking out relational/equality ops from "other binary ops".  This would leave us with the following groups:

1. ```*  /  %  +  -  <<  >>  &  ^  |```
2. ```<  >  <=  >=  is  as  ==  !=```
3. ```&& ||  ??```

However, looking into things more, the way the feature works would still be unsatisfactory.  Tehre is a general design problem we have here in terms of what it means to specify parentheses preferences for *any* operators.  Consider a preference that says you want parentheses for relational operators like ```==```.  Does that mean you want these parens:

```(a == b) || (c == d)```

i.e. do you want the parens *around* the binary expression with that op?  or do you want these parens:

```(a - b) == (c - d)```

i.e. do you want the parens around the *expressions* that constitute the binary expression with that op.  Basically, do you want parens on the *inside* of the binary op, or hte outside?   

this is confusing because both are valid perspectives.  So for the former, one person might think these are hte parens on the inside of the ```||``` and another may think they parens on the outside of the ```==``` expressions. 

What was interesting about this exploration was that as we looked at examples, one thing that seemed clear was that parens inside these groups made a lot of sense, but parens *between* the groups seemed less necessary.  

In other words, in both those above examples, parentheses seem very unnecessary no matter what the perspective.  There is little confusion around 

```(a == b) || (c == d)``` vs ```a == b || c == d```   or even
```(a - b) == (c - d)``` vs ```a - b == c - d```

The confusion we see is usually within the groups.  i.e. ```a || b && c ?? d``` is fairly confusing, but ```a < b && c == d``` far less so.  People have a general intuition that the logical operators are weaker than the equality/relational ops, and that hte equality/relational ops are weaker than the arithmetic ops.  However, within those groups there can be much less certainty about where things stand.

Based on that, we think a reasonable new approach to try is to have the above three groups, but to only consider parens *within* nested binary ops within that same group.  i.e. if there's a large complex expression  of hte same type: ```a op1 b op2 c op3 d``` we could generate any of 

```(a op1 b) op2 c op3 d```
```(a op1 b op2 c) op3 d```
```a op1 (b op2 c) op3 d```
```a op1 (b op2 c op3 d)```
```a op1 b op2 (c op3 d)```
```(a op1 b) op2 (c op3 d)```

We would not generate:

```(a op1 b op2 c op3 d)```

CyrusNajmabadi commented:
@sharwell I'm guessing you're  busy, but have you had any time to look into the viability of this approach?  Thanks!

sharwell commented:
@CyrusNajmabadi Where are we going to implement the restriction that parentheses are only considered within groups of the same type? Is that in #26710?

sharwell commented:
@jinujoseph Requesting approval to merge. This change implements the original approved design for this feature. The pending design change is implemented in #26710 should we decide to adopt it.

CyrusNajmabadi commented:
@sharwell Works for me.

jinujoseph commented:
Approved to merge for 15.8.preview3
